### PR TITLE
Fix/ Sprint number, date and points.

### PR DESCRIPTION
**Pull request by kostadriano**
Fix bug of number, date and points of sprints on board V2 [#24016](https://central.cm42.io/projects/central-board-v2#story-24016)

Before:
image

After:
image

### CODE REVIEW

File: app/assets/javascripts/models/beta/iteration.js [hunk 1]
```diff
@@ -13,11 +13,25 @@ const getIterationForDate = (date, project) => {
     : iterationQuantity + 1;
 };
 
+const getNextIterationNumber = (sprints) => {
+  const lastSprint = sprints[sprints.length - 1];
```
Comment (nolleto) [inline app/assets/javascripts/models/beta/iteration.js]: I've checked that this repository has [underscorejs](https://underscorejs.org/#last) so, you could use the `last` function here.
```js
import { last } from 'underscore';
...
const lastSprint = last(sprints);

File: app/assets/javascripts/models/beta/iteration.js [hunk 2]
```diff
@@ -13,11 +13,25 @@ const getIterationForDate = (date, project) => {
...
+
+  return lastSprint ? lastSprint.number + 1 : 1;
+}
+
 export const getDateForIterationNumber = (iterationNumber, project) => {
-  return moment(new Date(project.startDate))
-    .startOf("isoWeek")
-    .add(iterationNumber, "weeks")
-    .format("ddd MMM Do Y");
+  let startDate;
```
Comment (nolleto) [inline app/assets/javascripts/models/beta/iteration.js]: I think you can break this function in two:
```js
const getStartDate = (iterationNumber, project) => {
  const momentDate = moment(project.startDate);

  if (iterationNumber === 1) {
    return moment(project.startDate);
  } else {
    const missingDaysFromFirstSprint = (momentDate.isoWeekday() - project.iterationStartDay) % 7;
    const iterationDays = (iterationNumber - 1) * (project.iterationLength * 7);

    return (momentDate.subtract(missingDaysFromFirstSprint, 'days')).add(iterationDays, 'days');
  }
}

export const getDateForIterationNumber = (iterationNumber, project) => {
  return getStartDate(iterationNumber, project).format("ddd MMM Do Y");
};
```

File: app/assets/javascripts/selectors/columns.js [hunk 3]
```diff
@@ -20,7 +20,12 @@ export const getColumns = createSelector(
           story => Column.isBacklog(story, project))
         );
 
-        return groupStoriesInSprints(orderedStories, project);
+        const lastPastIteration = pastIterations[pastIterations.length - 1];
```
Comment (nolleto) [inline app/assets/javascripts/selectors/columns.js]:
```js
import { last } from 'underscore';
...
const lastPastIteration = last(pastIterations);
```

File: app/assets/javascripts/models/beta/iteration.js (hunk 1 above)
Comment (kostadriano) [reply app/assets/javascripts/models/beta/iteration.js]: uuu, fancy :smile:

File: app/assets/javascripts/models/beta/iteration.js (hunk 2 above)
Comment (talyssonoc) [reply app/assets/javascripts/models/beta/iteration.js]: I agree. Also you dont need `else` when you have a `return` inside the `if` case 😄

File: app/assets/javascripts/models/beta/iteration.js [hunk 4]
```diff
@@ -30,13 +44,13 @@ export const getIterationForStory = (story, project) => {
 };
 
 const createSprint = (sprintNumber = 0, startDate = 0, isFiller = false, velocity) => ({
-  number: sprintNumber + 1,
+  number: sprintNumber,
   startDate: startDate,
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/iteration.js]: We can improve it by a bit (now that our Babel supports it) using the shorthand notation:

```suggestion
  startDate,
```

File: app/assets/javascripts/models/beta/iteration.js [hunk 5]
```diff
@@ -30,13 +44,13 @@ export const getIterationForStory = (story, project) => {
...
   points: 0,
   completedPoints: null,
   stories: [],
   isFiller: isFiller,
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/iteration.js]: Same here about using the shorthand notation

```suggestion
  isFiller,
```

File: app/assets/javascripts/models/beta/iteration.js (hunk 2 above)
Comment (kostadriano) [reply app/assets/javascripts/models/beta/iteration.js]: sure :smile:

File: app/assets/javascripts/selectors/done.js [hunk 6]
```diff
@@ -1,13 +1,12 @@
...
-    startDate: moment(sprint.startDate).format("ddd MMM Do Y"),
-    endDate: moment(sprint.endDate).format("ddd MMM Do Y"),
-    stories: [],
-  }));
-
-  return sprints;
-};
+export const mountPastIterations = (pastIterations) =>
+  pastIterations.map(sprint =>
+    ({
```
Comment (talyssonoc) [inline app/assets/javascripts/selectors/done.js]: Can we put it in the same line as the arrow? It follows the standard we use for the rest of the app.

File: app/assets/javascripts/selectors/done.js [hunk 7]
```diff
@@ -1,13 +1,12 @@
...
+      ...sprint,
+      number: sprint.iterationNumber,
+      startDate: moment(sprint.startDate).format("ddd MMM Do Y"),
```
Comment (talyssonoc) [inline app/assets/javascripts/selectors/done.js]: Is this `format` considering I18n?

File: app/assets/javascripts/models/beta/iteration.js [hunk 8]
```diff
@@ -13,12 +14,27 @@ const getIterationForDate = (date, project) => {
...
-    .format("ddd MMM Do Y");
-};
+const getNextIterationNumber = (sprints) => {
+  const lastSprint = last(sprints);
+
+  return lastSprint ? lastSprint.number + 1 : 1;
+}
+
+const getStartDate = (iterationNumber, project) => {
+  const momentDate = moment(project.startDate);
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/iteration.js]: What do you think about renaming it to `projectStartDate`?

File: app/assets/javascripts/models/beta/iteration.js [hunk 9]
```diff
@@ -13,12 +14,27 @@ const getIterationForDate = (date, project) => {
...
+
+  if (iterationNumber === 1) {
+    return moment(project.startDate);
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/iteration.js]: We can return the variable from the first line of this function here.

File: app/assets/javascripts/models/beta/iteration.js [hunk 10]
```diff
@@ -13,12 +14,27 @@ const getIterationForDate = (date, project) => {
...
+  }
+
+  const missingDaysFromFirstSprint = (momentDate.isoWeekday() - project.iterationStartDay) % 7;
```
Comment (talyssonoc) [inline app/assets/javascripts/models/beta/iteration.js]: What do you think about extracting this logic to somewhere else to make it more explicit?

**Pull request merged**